File: algorithms/015/Solution.java
import java.util.*;

public class Solution {

    public List<List<Integer>> threeSum(int[] nums) {
        return nsum(0, 3, nums);
    }

    public List<List<Integer>> nsum(int target, int n, int[] nums) {
        Arrays.sort(nums);
        return nsum(target, n, nums, 0, nums.length);
    }

    public List<List<Integer>> nsum(int target, int n, int[] nums, int start, int end) {
        int size = end - start;
        if (size < n) {
            return new ArrayList<>();
        }

        if (n == 1) {
            List<List<Integer>> res = new ArrayList<>();
            int targetIndex = Arrays.binarySearch(nums, start, end, target);
            if (targetIndex >= 0) {
                LinkedList<Integer> l = new LinkedList<>();
                l.add(target);
                res.add(l);
            }
            return res;
        }

        List<List<Integer>> res = new ArrayList<>();
        for (int i = start; i < end - n + 1; i++) {
            if (i > start && nums[i - 1] == nums[i])
                continue;

            int num = nums[i];
            List<List<Integer>> subRes = nsum(target - num, n - 1, nums, i + 1, end);

            for (List<Integer> l : subRes) {
                LinkedList<Integer> ll = (LinkedList<Integer>) l;
                ll.addFirst(num);
                res.add(ll);
            }
        }
        return res;
    }
}

File: algorithms/015/Solution.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    let res = [];
    nums = nums.sort((a, b) => a - b);
    for (let i = 0; i < nums.length - 2; i++) {
        if (nums[i - 1] !== undefined && nums[i - 1] === nums[i]) {
            continue;
        }
        let num = nums[i];
        let m = i + 1;
        let n = nums.length - 1;
        while (m < n) {
            if (nums[m] + nums[n] + num === 0) {
                res.push([num, nums[m], nums[n]]);
                while (m < n && nums[m] === nums[m + 1]) {
                    m++;
                }
                while (m < n && nums[n] === nums[n - 1]) {
                    n--;
                }
                m++;
                n--;
            } else if (nums[m] + nums[n] + num > 0) {
                n--;
            } else {
                m++;
            }
        }
    }
    return res;
};
